File: 39-combination-sum/39-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans=[[] for i in range(target)]
        for i in range(len(candidates)):
            for j in range(target):
                if j+1-candidates[i]==0:
                    ans[j].append({j+1:1})
                if j-candidates[i]>=0:
                    for x in ans[j-candidates[i]]:
                        tmp=dict()
                        for y in x:
                            if y==candidates[i]:
                                tmp[y]=x[y]+1
                            else:
                                tmp[y]=x[y]
                        if candidates[i] not in tmp:
                            tmp[candidates[i]]=1
                        ans[j].append(tmp)
        a=[]
        for x in ans[target-1]:
            tmp=[]
            for y in x:
                for i in range(x[y]):
                    tmp.append(y)
            a.append(tmp)
        return a

File: 39-combination-sum/39-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<map<int,int>>> ans(target);
        for(int i=0;i<candidates.size();i+=1){
            for(int j=0;j<target;j+=1){
                if(j+1-candidates[i]==0){
                    map<int,int> m;
                    m[j+1]=1;
                    ans[j].push_back(m);
                }
                if(j>=candidates[i]){
                    for(auto x: ans[j-candidates[i]]){
                        map<int,int> tmp;
                        for(auto y:x){
                            if(y.first==candidates[i]){
                                tmp[y.first]=y.second+1;
                            }else{   
                                 tmp[y.first]=y.second;
                             }
                        }
                        if(tmp.find(candidates[i])==tmp.end()){
                            tmp[candidates[i]]=1;
                            }
                        ans[j].push_back(tmp);
                    }
                }
            }
        }
        vector<vector<int>> a;
        for(auto x: ans[target-1]){
            vector<int> tmp;
            for(auto y:x){
                for(int i=0;i<y.second;i+=1)
                    tmp.push_back(y.first);
            }
            a.push_back(tmp);
        }
        return a;
    }
};